File: hphp/hack/src/rupro/lib/utils/core/ident.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.
use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct Ident(u64);

impl From<u64> for Ident {
    fn from(x: u64) -> Self {
        Self(x)
    }
}

#[derive(Debug, Clone)]
pub struct IdentGen {
    next: Rc<RefCell<u64>>,
}

impl IdentGen {
    pub fn new() -> Self {
        Self {
            next: Rc::new(RefCell::new(1)),
        }
    }

    pub fn make(&self) -> Ident {
        let mut r = self.next.borrow_mut();
        let v = *r;
        *r += 1;
        Ident(v)
    }
}

File: hphp/hack/src/heap/ident.ml
(*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Hh_prelude

external hh_counter_next : unit -> int = "hh_counter_next"

type t = int [@@deriving eq, hash, ord, show]

let track_names = ref false

let trace = ref IMap.empty

let set_name id name = trace := IMap.add id name !trace

let tmp () =
  let res = hh_counter_next () in
  if !track_names then set_name res ("__tmp" ^ string_of_int res);
  res

let to_string x =
  match IMap.find_opt x !trace with
  | Some res -> res
  | None -> "v" ^ string_of_int x

let make name =
  let res = hh_counter_next () in
  if !track_names then set_name res name;
  res
